Create vulnerability-handling doc (#33)
jupyter/security: docs/vulnerability-handling.md

## Coordinated Disclosures

Project Jupyter follows a [coordinated disclosure](https://cheatsheetseries.owasp.org/cheatsheets/Vulnerability_Disclosure_Cheat_Sheet.html#responsible-or-coordinated-disclosure)
model where the initial report and remediation are handled privately, but the completion description is made public once a patch
is available. Project Jupyter will disclose known vulnerabilities within 90 days by default, whether a patch is available or not.

## Acknowledgement

Project Jupyter will work to ensure that security researchers, developers, users, or others who identify and report vulnerabilities
within Project Jupyter software receive acknowledgement for their contribution. 

## Vulnerability Triage & Remediation Process

This section describes the process used by Project Jupyter to track, remediate, and disclose reported vulnerabilities. 
This description is both a reference for the Jupyter Community and a guideline for contributors.

### Roles

This process defines these roles:
- **Reporter** The individual(s) who report the vulnerability 
- **Coordinator** A Project Jupyter contributor who facilitates the tracking of the vulnerability through this process

> - **Developer** One or more Project Jupyter developers who work on remediating the vulnerability

Author: Should these roles be separate individuals? I.e., could someone do two or all of these roles?

I don't think it's mandatory that these be separate people. A developer should be allowed to fix a vulnerability they reported, for example.

Reviewer: ```suggestion
This process defines these roles:
- **Reporter** The individual(s) who report the vulnerability 
- **Coordinator** A Project Jupyter contributor who facilitates the tracking of the vulnerability through this process
- **Developer** One or more Project Jupyter developers who work on remediating the vulnerability

For the purpose of this document these roles are distinct, in practice, some of these roles may be handled by the same individual. However, the roles should be covered by a minimum of two separate individuals. For example, a Reporter may also fill the Developer role and create the remediation, in this case the Coordinator should be a separate individual.
```

### Process

> The role responsible for each step is noted at the beginning.

Author: Might be good to have an expectation for a communication in response to a report: an initial response within X days/hours and a follow-up within Y days/hours of filing, for example.

Who will own the creation of mechanisms to keep reports on track for our process?

- Upon receipt of the initial report:
  - **Coordinator**: Respond to the reported and acknowledge receipt of the report 
  - **Coordinator**: Open an issue in the private GitHub repository used for tracking vulnerabilities across projects

>   - **Coordinator**: Assign the issue to one or more **Developers** to confirm the report

Author: ```suggestion
  - **Coordinator**: Review the issue for completeness and suitability (triage). If more information is needed, follow up with the Reporter.
```

  - **Coordinator**: Notify the reporter
- If the vulnerability is confirmed, within the relevant repositories:
  - **Coordinator**: Open a draft [GitHub Security Advisory](https://docs.github.com/en/code-security/repository-security-advisories/about-github-security-advisories-for-repositories#about-github-security-advisories)
    - Include relevant but sanitized details in the top level comment, which becomes public
    - Sensitive details and reproductions go in the comments on the draft advisory, which are not public
    - **Coordinator**: Add relevant people to the advisory - *note* we need an official list of GitHub handles here since teams don't work across orgs
    - **Coordinator**: Request a [CVE](https://docs.github.com/en/code-security/repository-security-advisories/about-github-security-advisories-for-repositories#cve-identification-numbers) from GitHub.  This is private until the advisory is published.
    - **Developer**: Work on the [vulnerability fix PR](https://docs.github.com/en/code-security/repository-security-advisories/collaborating-in-a-temporary-private-fork-to-resolve-a-repository-security-vulnerability#creating-a-temporary-private-fork)

>     - Once the fix is agreed upon, wait for the CVE number to be issued.

Reviewer 2: I'm not sure if this is in scope here or if it belongs elsewhere: It'd be good to add a warning that GitHub CI workflows won't run on security forks, so reviewers must test manually to avoid a broken CI when the patch is merged to the public repo.

Reviewer: I think it's very appropriate to mention here, since this is meant to capture the process for developers woking on the remediation. Better to have some extra notes than miss something.

Along with the lack of CI, some vulnerabilities may involve multiple private security forks across different GitHub org. So it's worth mentioning both the need for manual testing, and for testing with any other parallel remediations.

Reviewer 3: We added a note at the bottom, we can rene later.

within Project Jupyter software receive acknowledgement for their contribution. 

## Vulnerability Triage & Remediation Process

This section describes the process used by Project Jupyter to track, remediate, and disclose reported vulnerabilities. 
This description is both a reference for the Jupyter Community and a guideline for contributors.

### Roles

This process defines these roles:
- **Reporter** The individual(s) who report the vulnerability 
- **Coordinator** A Project Jupyter contributor who facilitates the tracking of the vulnerability through this process
- **Developer** One or more Project Jupyter developers who work on remediating the vulnerability

For the purpose of this document these roles are distinct, in practice, some of these roles may be handled by the same individual. However, the roles should be covered by a minimum of two separate individuals. For example, a Reporter may also fill the Developer role and create the remediation, in this case the Coordinator should be a separate individual.

### Process

The role responsible for each step is noted at the beginning.

- Upon receipt of the initial report:

>   - **Coordinator**: Respond to the reported and acknowledge receipt of the report 

Reviewer 3: ```suggestion
  - **Coordinator**: Respond to the reported and acknowledge receipt of the report in the timeframe given in the "Reporting Vulnerabilities" section.
```

  - **Coordinator**: Open an issue in the private GitHub repository used for tracking vulnerabilities across projects
  - **Coordinator**: Assign the issue to one or more **Developers** to confirm the report

>   - **Developer**: Attempt to replicate the reported vulnerability. Request more information from the **Reporter** if necessary. Update the tracking issue with the results.

Reviewer: ```suggestion
```

- If the vulnerability cannot be confirmed:
  - **Coordinator**: Close the issue
  - **Coordinator**: Notify the reporter

> - If the vulnerability is confirmed, within the relevant repositories:

Reviewer 3: :+1:

  - **Coordinator**: Open a draft [GitHub Security Advisory](https://docs.github.com/en/code-security/repository-security-advisories/about-github-security-advisories-for-repositories#about-github-security-advisories)
    - Include relevant but sanitized details in the top level comment, which becomes public
    - Sensitive details and reproductions go in the comments on the draft advisory, which are not public

>     - **Coordinator**: Add relevant people to the advisory - *note* we need an official list of GitHub handles here since teams don't work across orgs

Reviewer: Clarify who should be added, like the Reporter, specific developers, etc. Also consider anyone maintaining documentation.

>     - **Coordinator**: Request a [CVE](https://docs.github.com/en/code-security/repository-security-advisories/about-github-security-advisories-for-repositories#cve-identification-numbers) from GitHub.  This is private until the advisory is published.

Reviewer 3: ```suggestion
    - **Coordinator**/**Developer**: Request a [CVE](https://docs.github.com/en/code-security/repository-security-advisories/about-github-security-advisories-for-repositories#cve-identification-numbers) from GitHub. The CVE Number will be private until the advisory is published.
```

>     - **Coordinator**: Request a [CVE](https://docs.github.com/en/code-security/repository-security-advisories/about-github-security-advisories-for-repositories#cve-identification-numbers) from GitHub.  This is private until the advisory is published.

Reviewer: ```suggestion
    - **Developer**: Attempt to replicate the reported vulnerability. Request more information from the **Reporter** if necessary.
    - **Coordinator**: Request a [CVE](https://docs.github.com/en/code-security/repository-security-advisories/about-github-security-advisories-for-repositories#cve-identification-numbers) from GitHub.  This is private until the advisory is published.
```

    - **Developer**: Work on the [vulnerability fix PR](https://docs.github.com/en/code-security/repository-security-advisories/collaborating-in-a-temporary-private-fork-to-resolve-a-repository-security-vulnerability#creating-a-temporary-private-fork)
    - Once the fix is agreed upon, wait for the CVE number to be issued.

>     - **Coordinator**: Notify the **Reporter** of the CVE number

Reviewer 3: ```suggestion
    - **Coordinator**: Notify the **Reporter** of the CVE number. As the CVE number does not contain sensitive informations, it can also be communicated to downstream vendors with the affected software name (without revealing version numbers).
```

    - **Developer & Coordinator**: Decide on release and announcement dates and post them the draft advisory.  These are typically at least a week apart to allow administrators to deploy fixes.
  - **Coordinator**: Post the release and announcement dates on the Jupyter Security [Mailing List]([security@ipython.org](mailto:security@ipython.org))
  - **Developer**: Merge the security fix PR
  - **Developer**: Make a release to PyPI and/or npm with no announcement or change log
  - **Coordinator**: Publish the security advisory on the announcement date.  GitHub will post the CVE to the MITRE database
- **Coordinator**: Notify the **Reporter** of the releases
- **Coordinator**: Close the issue in the tracking repository

> 

Reviewer: ```suggestion
### Note to Developers

Be aware that GitHub CI workflows won't run on security forks, so reviewers must test manually to avoid a broken CI when the patch is merged to the public repo. Also, vulnerabilities may involve multiple private security forks across different GitHub organizations. This may require additional manual steps to include those private forks.
```
